codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/BasicUserAgentFeatures.java

```diff
@@ -0,0 +1,88 @@
+package software.amazon.smithy.aws.go.codegen.customization;
+
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+import static software.amazon.smithy.go.codegen.SymbolUtils.buildPackageSymbol;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiPredicate;
+import software.amazon.smithy.aws.go.codegen.AwsGoDependency;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.go.codegen.GoCodegenContext;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SmithyGoTypes;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.protocol.traits.Rpcv2CborTrait;
+
+/**
+ * Adds user agent tracking for basic features - i.e. simple model-based ones that do not require any additional in-code
+ * checks, such as a particular protocol.
+ */
+public class BasicUserAgentFeatures implements GoIntegration {
+    private static final List<Feature> FEATURES = List.of(
+           new Feature("ProtocolRPCV2CBOR", (model, service) -> service.hasTrait(Rpcv2CborTrait.class))
+    );
+
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return FEATURES.stream().map(Feature::getPlugin).toList();
+    }
+
+    @Override
+    public void writeAdditionalFiles(GoCodegenContext ctx) {
+        var model = ctx.model();
+        var service = ctx.settings().getService(model);
+        ctx.writerDelegator().useFileWriter("api_client.go", ctx.settings().getModuleName(),
+                GoWriter.ChainWritable.of(
+                        FEATURES.stream()
+                                .filter(it -> it.servicePredicate.test(model, service))
+                                .map(Feature::getAddMiddleware)
+                                .toList()
+                ).compose());
+    }
+
+    private static final class Feature {
```

Comment:
While this is named `Feature` it's really only for CBOR metrics due to the `getAddMiddleware()`. We should either make the CBOR metrics extend from this or rename this

Reply:
What about the `getAddMiddleware` constrains it to CBOR? The feature ID variant is passed in.

Reply:
Yeah, my point is that the class name is generic but the functionality is not, so we should either make it generic or name it accordingly

Reply:
The functionality of `Feature` _is_ generic though - it's just that CBOR is the only thing using it. You could add any other static feature to the `FEATURES` list in the future.

e.g.
```
    private static final List<Feature> FEATURES = List.of(
           new Feature("ProtocolRPCV2CBOR", (model, service) -> service.hasTrait(Rpcv2CborTrait.class)),
           new Feature("ProtocolSparrowhawk", (model, service) -> etc())
    );

Reply:
OK I'm dumb, I kept reading the functionality of `getAddMiddleware` as specific to CBOR since the code always calls the same middleware, but this is exactly what we want since this middleware is generic